rclcpp_action/include/rclcpp_action/server.hpp

```diff
@@ -31,65 +36,428 @@ namespace rclcpp_action
+   *  communicate.
+   * \param[in] options options to pass to the underlying `rcl_action_server_t`.
+   * \param[in] handle_goal a callback that decides if a goal should be accepted or rejected.
+   * \param[in] handle_cancel a callback that decides if a goal should be attemted to be canceled.
+   *  The return from this callback only indicates if the server will try to cancel a goal.
+   *  It does not indicate if the goal was actually canceled.
+   * \param[in] handle_accepted a callback that is called to give the user a handle to the goal.
+   *  execution.
+   */
   Server(
     rclcpp::node_interfaces::NodeBaseInterface::SharedPtr node_base,
+    rclcpp::node_interfaces::NodeClockInterface::SharedPtr node_clock,
     const std::string & name,
+    const rcl_action_server_options_t & options,
     GoalCallback handle_goal,
-    CancelCallback handle_cancel
+    CancelCallback handle_cancel,
+    AcceptedCallback handle_accepted
   )
   : ServerBase(
       node_base,
+      node_clock,
       name,
-      rosidl_typesupport_cpp::get_action_type_support_handle<ACTION>()),
+      rosidl_typesupport_cpp::get_action_type_support_handle<ACTION>(),
+      options),
     handle_goal_(handle_goal),
-    handle_cancel_(handle_cancel)
+    handle_cancel_(handle_cancel),
+    handle_accepted_(handle_accepted)
   {
-    // TODO(sloretz) what's the link that causes `handle_goal_` and `handle_cancel_` to be called?
   }
 
-  virtual ~Server()
+  virtual ~Server() = default;
+
+protected:
+  // -----------------------------------------------------
+  // API for communication between ServerBase and Server<>
+
+  /// \internal
+  std::pair<GoalResponse, std::shared_ptr<void>>
+  call_handle_goal_callback(std::array<uint8_t, 16> & uuid, std::shared_ptr<void> message) override
   {
+    // TODO(sloretz) update and remove assert when IDL pipeline allows nesting user's type
+    static_assert(
+      std::is_same<typename ACTION::Goal, typename ACTION::GoalRequestService::Request>::value,
+      "Assuming user fields were merged with goal request fields");
+    GoalResponse user_response = handle_goal_(
+      uuid, std::static_pointer_cast<typename ACTION::Goal>(message));
+
+    auto ros_response = std::make_shared<typename ACTION::GoalRequestService::Response>();
+    ros_response->accepted = GoalResponse::ACCEPT_AND_EXECUTE == user_response ||
+      GoalResponse::ACCEPT_AND_DEFER == user_response;
+    return std::make_pair(user_response, ros_response);
   }
 
+  /// \internal
+  std::pair<CancelResponse, std::shared_ptr<rcl_action_goal_handle_t>>
+  call_handle_cancel_callback(const std::array<uint8_t, 16> & uuid) override
+  {
+    CancelResponse resp = CancelResponse::REJECT;
+    std::shared_ptr<rcl_action_goal_handle_t> rcl_handle;
+    auto element = goal_handles_.find(uuid);
+    if (element != goal_handles_.end()) {
+      std::shared_ptr<ServerGoalHandle<ACTION>> goal_handle = element->second.lock();
+      if (goal_handle) {
```

Comment:
I don't think so. It would happen if the original `goal_handle` was destructed, so there's nothing left that care's about the cancel request.

Though maybe the `goal_handle` should try transition itself to `ABORTED` when it gets destructed without completing.